### PR TITLE
feat(helm): implement new index format

### LINKED ISSUE #1197: Alter schema of 'index.yaml'
The current index.yaml file looks like this:

```yaml
mariner-4.3.2:
  name: mariner-4.3.2
  url: localhost:8879/charts/mariner-4.3.2.tgz
  chartfile:
    name: mariner
    version: "4.3.2"
    description: Some description
    icon: https://example.com/path/to/icon.png
mariner-4.1.1:
  name: mariner-4.1.1
  url: localhost:8879/charts/mariner-4.1.1.tgz
  chartfile:
    name: mariner
    version: "4.3.2"
    description: Some description
    icon: https://example.com/path/to/icon.png
alpine-0.1.0:
  name: alpine-0.1.0
  url: localhost:8879/charts/alpine-0.1.0.tgz
  chartfile: ...
```

There have been ongoing discussions to alter this format before Alpha.5 releases. The goals for altering are:

- Fewer repeated fields
- Flatter chart metadata
- Easier structure for clients to process, where clients may be `helm` or something much simpler like `jq`.

Here's the proposed new version:

```yaml
mariner:
  - version: "4.3.2"
     url: localhost:8879/charts/mariner-4.3.2.tgz
     digest: "sha256:..."
     name: mariner
     description: Some description
     icon: https://example.com/path/to/icon.png
     # rest of Chart.yaml fields
  - version: "4.1.1"
    url: localhost:8879/charts/mariner-4.1.1.tgz
    digest: "sha256:..."
    name: mariner
    description: Some description
    icon: https://example.com/path/to/icon.png
    # rest of Chart.yaml fields
alpine:
  - version: 0.1.0
     url: localhost:8879/charts/alpine-0.1.0.tgz
     # ...
```

This presents the data in a more tree-like way:

```
CHART:
    VERSIONS[]
CHART:
    VERSIONS[]
```

Note that it does collapse the `chartfile` fields into the versions object.

When it comes to ordering the VERSIONS[] list, it should be ordered by version, descending. So the newest release is in VERSIONS[0]. Assessing the stability of a version is done by SemVer 2 rules.

### PR DESCRIPTION
**Pull request by technosophos**
This implements a new index file format for repository indices. It also
implements a new format for requirements.yaml.

**Breaking change:** This will break all previous versions of Helm, and will
impact helm search, repo, serve, and fetch functions. Old chart indices will also be broken, since the `index.yaml` format is different.

Closes #1197

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/helm/1257)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -62,7 +62,7 @@ test: test-unit
 
 .PHONY: test-unit
 test-unit:
-	$(GO) test $(GOFLAGS) -run $(TESTS) $(PKG) $(TESTFLAGS)
+	HELM_HOME=/no/such/dir $(GO) test $(GOFLAGS) -run $(TESTS) $(PKG) $(TESTFLAGS)
```
Comment (adamreese) [inline Makefile]: =/

Comment (technosophos) [inline Makefile]: I can pull that if you object. But I needed a quick way to make it clear to test authors that their tests were munging data in $HELM_HOME.

File: cmd/helm/init_test.go [hunk 2]
```diff
@@ -59,9 +63,3 @@ func TestEnsureHome(t *testing.T) {
 		t.Errorf("%s should not be a directory", fi)
 	}
 }
-
-func createTmpHome() string {
```
Comment (adamreese) [inline cmd/helm/init_test.go]: how did this ever work?

Comment (technosophos) [inline cmd/helm/init_test.go]: I felt bad laughing when I read that one.

**Comment by technosophos** [comment]
See #1265 for a follow-up issue on this one.

**Comment by technosophos** [comment]
And here's a repo for testing with the new index format: https://github.com/technosophos/tscharts

**Comment by technosophos** [comment]
To test:

```
$ make build
$ helm init --client-only # You should see a message about updating the repositories file
$ helm update  # This should give warnings about index files being in old format
$ helm repo add tscharts https://technosophos.github.io/tscharts
$ helm update # The tscharts should get updated, but you should still see warnings
$ helm search al # should show tscharts/alpine, maybe warnings for other repos
$ helm fetch tscharts/alpine-0.1.0 # should work
$ helm install tscharts/alpine-0.1.0  # should work
```

To test creating a new repo:

```
$ mkdir myrepo
$ cd myrepo
$ helm create foobar
$ helm package foobar
$ cd ..
$ helm repo index myrepo --url https://example.com
```

You should be able to visually inspect the index.yaml for the new format

You should be able to start a local repo with:
```
$ helm serve
```

Then from another terminal, you can `helm fetch local/foobar-0.1.0` .

File: cmd/helm/init.go [hunk 3]
```diff
@@ -101,9 +107,8 @@ func requireHome() error {
 // ensureHome checks to see if $HELM_HOME exists
 //
 // If $HELM_HOME does not exist, this function will create it.
-func ensureHome() error {
-	configDirectories := []string{homePath(), repositoryDirectory(), cacheDirectory(), localRepoDirectory()}
-
+func ensureHome(home helmpath.Home) error {
```
Comment (adamreese) [inline cmd/helm/init.go]: could accept an `io.writer`

File: pkg/repo/local.go [hunk 4]
```diff
@@ -23,21 +23,28 @@ import (
...
 
 var localRepoPath string
 
 // StartLocalRepo starts a web server and serves files from the given path
-func StartLocalRepo(path string) {
+func StartLocalRepo(path, address string) error {
+	if address == "" {
+		address = ":8879"
+	}
 	fmt.Println("Now serving you on localhost:8879...")
```
Comment (adamreese) [inline pkg/repo/local.go]: library code printing to stdout 0_o

File: cmd/helm/package.go [hunk 5]
```diff
@@ -119,10 +122,11 @@ func (p *packageCmd) run(cmd *cobra.Command, args []string) error {
 	// Save to $HELM_HOME/local directory. This is second, because we don't want
 	// the case where we saved here, but didn't save to the default destination.
 	if p.save {
-		if err := repo.AddChartToLocalRepo(ch, localRepoDirectory()); err != nil {
+		lr := p.home.LocalRepository()
+		if err := repo.AddChartToLocalRepo(ch, lr); err != nil {
 			return err
 		} else if flagDebug {
-			cmd.Printf("Saved %s to %s\n", name, localRepoDirectory())
+			cmd.Printf("Saved %s to %s\n", name, lr)
```
Comment (technosophos) [inline cmd/helm/package.go]: Well that's annoying. ;-). I'll switch to fmt.Fprintf

**Comment by technosophos** [comment]
@adamreese I think I got all your concerns, plus I switched some of the `helm_test` to use `t.Logf` instead of `fmt.Printf`.

**Pull request merged**